### PR TITLE
Feature/catalog browsing

### PR DESCRIPTION
**Pull request by AndriiSameliuk**
Implement catalog browsing:
1. Implement API for the hierarchy of categories.
2. Implement `Product` and `Category` entities.
3. Implement API for searching.

### CODE REVIEW

**Comment by AndriiSameliuk** [comment]
@BohdanDudkaTD, PTAL.

File: src/test/java/io/javaclasses/hookahpalace/service/CategoryTest.java [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    @ParameterizedTest
+    @MethodSource("addInvalidCategory")
+    void testCategoryFindingByIdWithInvalidProduct(Category category, String assertMessage) throws ValidationException {
+        Assertions.assertThrows(ValidationException.class, () -> categoryService.findById(category.id()), assertMessage);
+    }
+
+    private static Stream<Arguments> addInvalidCategory() {
+        return Stream.of(
+                Arguments.of(new Category("2", "Hookah1", "1"),
+                        "Exception due to incorrect parent id of category wasn't thrown.")
```
Comment (BohdanDudkaTD) [inline src/test/java/io/javaclasses/hookahpalace/service/CategoryTest.java]: Is it a good idea to have one assertion message for 4 test cases?
Please, create a separate source method for each invalid test case.
Also, provide more detailed assertion messages.

File: src/test/java/io/javaclasses/hookahpalace/service/CategoryTest.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+    @ParameterizedTest
+    @MethodSource("addInvalidCategory")
+    void testCategoryDeletingWithInvalidProduct(Category category, String assertMessage) throws ValidationException {
+        Assertions.assertThrows(ValidationException.class, () -> categoryService.delete(category.id()), assertMessage);
+    }
+
+    @ParameterizedTest
+    @MethodSource("addInvalidCategory")
+    void testCategoryUpdatingWithInvalidProduct(Category category, String assertMessage) throws ValidationException {
```
Comment (BohdanDudkaTD) [inline src/test/java/io/javaclasses/hookahpalace/service/CategoryTest.java]: Let's have more test cases for updating, deleting, searching and creating operations.
For example, a client tries to update a category that does not exist, update it with incorrect data, find the category which does not exist, create category, which already exists, etc.

File: src/test/java/io/javaclasses/hookahpalace/service/CategoryTest.java [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+                        "Exception due to incorrect parent id of category wasn't thrown.")
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("addCategoriesForGettingAllProductInside")
+    void testGettingAllProductsInCategory(Category category, List<Product> expectedList, String assetMessage)
+            throws ValidationException {
+        List<Product> actualList = categoryService.findAllProductByCategory(category);
+        Assertions.assertArrayEquals(expectedList.toArray(), actualList.toArray(), assetMessage);
```
Comment (BohdanDudkaTD) [inline src/test/java/io/javaclasses/hookahpalace/service/CategoryTest.java]: This test will always fail as the `equals` method is not overridden in the `Product` entity.
Please, see [this article about list testing](https://www.mkyong.com/unittest/junit-how-to-test-a-list/) (chapter 3)
image

File: src/test/java/io/javaclasses/hookahpalace/service/CategoryTest.java [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+                        "Getting all products inside category was failed.")
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("addCategoriesForGettingAllChildCategory")
+    void testGettingAllChildCategory(Category category, List<Category> expectedList, String assetMessage)
+            throws ValidationException {
+        List<Category> actualList = categoryService.findAllChildrenOfCategory(category);
+        Assertions.assertArrayEquals(expectedList.toArray(), actualList.toArray(), assetMessage);
```
Comment (BohdanDudkaTD) [inline src/test/java/io/javaclasses/hookahpalace/service/CategoryTest.java]: Same situation with the `equals` method.

File: src/test/java/io/javaclasses/hookahpalace/service/ProductTest.java [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        Assertions.assertThrows(ValidationException.class, () -> productService.update(product), assertMessage);
+    }
+
+    @ParameterizedTest
+    @MethodSource("addInvalidProducts")
+    void testProductFindingByIdWithInvalidProduct(Product product, String assertMessage) throws ValidationException {
+        Assertions.assertThrows(ValidationException.class, () -> productService.findById(product.id()), assertMessage);
+    }
+
+    private static Stream<Arguments> addInvalidProducts() {
```
Comment (BohdanDudkaTD) [inline src/test/java/io/javaclasses/hookahpalace/service/ProductTest.java]: Please, provide more test cases.

File: src/test/java/io/javaclasses/hookahpalace/service/SerchServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+    @Test
+    void testSearchServiceForSearchingProducts() throws ValidationException {
+        SearchResult searchResult = searchService.search("Something");
+        Product product = new Product("1",
+                "Something",
+                100,
+                new File("testimage/correctJpg.jpg"),
+                new Category("1", "Category").id());
+        Assertions.assertTrue(searchResult.products().contains(product));
```
Comment (BohdanDudkaTD) [inline src/test/java/io/javaclasses/hookahpalace/service/SerchServiceTest.java]: Please, provide an assertion message.
Also, this test will fail due to a lack of an `equals` method.

File: src/test/java/io/javaclasses/hookahpalace/service/SerchServiceTest.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    }
+
+    @Test
+    void testSearchServiceForSearchingCategories() throws ValidationException {
+        SearchResult searchResult = searchService.search("Something");
+        Category category = new Category("1", "Something");
+        Assertions.assertTrue(searchResult.categories().contains(category));
```
Comment (BohdanDudkaTD) [inline src/test/java/io/javaclasses/hookahpalace/service/SerchServiceTest.java]: Please, provide an assertion message.

File: src/main/java/io/javaclasses/hookahpalace/service/CategoryService.java [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+
+/**
+ * Makes queries to a database for executing CRUD operations on categories.
+ */
+public class CategoryService implements Repository<Category> {
+
+    /**
+     * Makes a query to a database for creating category.
+     *
+     * @param category {@link Category}
```
Comment (BohdanDudkaTD) [inline src/main/java/io/javaclasses/hookahpalace/service/CategoryService.java]: Please, describe the `category` parameter.

File: src/test/java/io/javaclasses/hookahpalace/service/CategoryServiceTest.java [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+    @MethodSource("addCategories")
+    void testCategoryUpdating(Category category) throws ValidationException {
+        Assertions.assertEquals(category, categoryService.update(category), "Updating of category was failed.");
+    }
+
+    @ParameterizedTest
+    @MethodSource("addCategories")
+    void testCategoryFindingById(Category category) throws ValidationException {
+        Category actualCategory = categoryService.findById(category.id());
+        Assertions.assertEquals(category.id(), actualCategory.id(), "Finding category by id was failed.");
```
Comment (BohdanDudkaTD) [inline src/test/java/io/javaclasses/hookahpalace/service/CategoryServiceTest.java]: There is no need to compare ids as `equals` method do this.

File: src/test/java/io/javaclasses/hookahpalace/service/ProductServiceTest.java [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+    @MethodSource("addProducts")
+    void testProductUpdating(Product product) throws ValidationException {
+        Assertions.assertEquals(product, productService.update(product), "Updating of product was failed.");
+    }
+
+    @ParameterizedTest
+    @MethodSource("addProducts")
+    void testProductFindingById(Product product) throws ValidationException {
+        Product actualProduct = productService.findById(product.id());
+        Assertions.assertEquals(product.id(), actualProduct.id(), "Finding product by id was failed.");
```
Comment (BohdanDudkaTD) [inline src/test/java/io/javaclasses/hookahpalace/service/ProductServiceTest.java]: Same here.

File: src/test/java/io/javaclasses/hookahpalace/service/SearchServiceTest.java [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+        Assertions.assertThrows(ValidationException.class, () -> {
+            SearchResult searchResult = searchService.search(keyword, productsAmount);
+            Category category = new Category("1", keyword);
+            searchResult.categories().contains(category);
+        }, assertMessage);
+    }
+
+    private static Stream<Arguments> addInvalidData() {
+        return Stream.of(
+                Arguments.of("Something", -1, "Exception due to incorrect products amount wasn't thrown."),
```
Comment (BohdanDudkaTD) [inline src/test/java/io/javaclasses/hookahpalace/service/SearchServiceTest.java]: "... incorrect amount of products to display ..."

**Pull request merged**